File: 1371-minimum-remove-to-make-valid-parentheses/1371-minimum-remove-to-make-valid-parentheses.js
/**
 * @param {string} s
 * @return {string}
 */
var minRemoveToMakeValid = function(s) {
    let stack = [];
    let invalidIndexes = new Set(); // 제거해야 할 인덱스 저장

    // 1. 스택을 사용해 유효하지 않은 괄호의 인덱스 추적
    for (let i = 0; i < s.length; i++) {
        let char = s[i];
        if (char === '(') {
            stack.push(i); // 여는 괄호의 인덱스 저장
        } else if (char === ')') {
            if (stack.length > 0) {
                stack.pop(); // 유효한 괄호 페어
            } else {
                invalidIndexes.add(i); // 닫는 괄호가 유효하지 않음
            }
        }
    }

    // 2. 스택에 남은 여는 괄호도 유효하지 않으므로 제거해야 함
    while (stack.length > 0) {
        invalidIndexes.add(stack.pop());
    }

    // 3. 유효한 문자만 결과에 추가
    let result = "";
    for (let i = 0; i < s.length; i++) {
        if (!invalidIndexes.has(i)) {
            result += s[i];
        }
    }

    return result;
};

File: 1371-minimum-remove-to-make-valid-parentheses/1371-minimum-remove-to-make-valid-parentheses.go
func minRemoveToMakeValid(s string) string {

    //TODO stack 을 정해두고 

    stack := make([]int, 0)
    set := make(map[int]bool)

    for i,v := range s {
        if v == ')' {
            if len(stack) > 0  {
                stack = stack[:len(stack) -1] //맨끝 요소 제거
            } else {
                // 없으면
                set[i] = true
            }
        } else if v == '(' {
                stack = append(stack,i)
            } 
    }

    for len(stack) > 0 {
        top := stack[len(stack) -1]
        set[top] = true //맵에 추가
        stack = stack[:len(stack)-1] //제거
    }

    result := ""
    for i,v := range s {
        if _,exists := set[i]; exists {
            continue
        }
        result += string(v)
    }


    //문자를 한글자 돌때마다 result 에 더하고
    // )를 만날때 top에 없으면, 그냥 스킵
    return result
}